modules/saku-policy-store/src/saku/dal_datalevin.clj

```diff
@@ -52,38 +43,77 @@
+    (impl-get-*-policies db' drns' :statement/identities)))
 
 
 (defmethod interface/get-identity-policies :datalevin [obj db-or-drns & [drns]]
   (let [[db' drns'] (if drns [db-or-drns drns] [(interface/db obj) db-or-drns])]
-    (impl-get-identity-policies db' drns')))
+    (impl-get-*-policies db' drns' :statement/resources)))
+
 
+(defn upsert-*-policies
+  [db-conn policies]
+  (let [tx-data (into []
+                  (mapcat (fn [{:keys [policy/drn] :as policy}]
+                            [[:db/retract [:policy/drn drn] :policy/statements]
+                             policy]))
+                  policies)]
+    (d/transact! db-conn tx-data)))
 
 (defmethod interface/upsert-resource-policies :datalevin [{:keys [db-conn] :as obj} policies]
   (schemas/assert* schemas/resource-policies policies)
-  (let [drns (map :policy/drn policies)
-        tx (reduce (fn [c {:keys [policy/drn] :as policy}]
-                     (conj c
-                           [:db/retract [:policy/drn drn] :policy/statements]
-                           policy))
-                   [] policies)
-        {:keys [db-after]} (d/transact! db-conn tx)]
-    (->> drns
-         (interface/get-resource-policies obj db-after))))
+  (let [{:keys [db-after]} (upsert-*-policies db-conn policies)]
+    (interface/get-resource-policies obj db-after (map :policy/drn policies))))
 
 
 (defmethod interface/upsert-identity-policies :datalevin [{:keys [db-conn] :as obj} policies]
   (schemas/assert* schemas/identity-policies policies)
-  (let [drns (map :policy/drn policies)
-        tx (reduce (fn [c {:keys [policy/drn] :as policy}]
-                     (conj c
-                           [:db/retract [:policy/drn drn] :policy/statements]
-                           policy))
-                   [] policies)
-        {:keys [db-after]}(d/transact! db-conn tx)]
-    (->> drns
-         (interface/get-identity-policies obj db-after))))
+  (let [{:keys [db-after]} (upsert-*-policies db-conn policies)]
+    (interface/get-identity-policies obj db-after (map :policy/drn policies))))
+
+(defn tx-fn-retract-statements-from-policy
+  [db {:keys [policy/drn statement-ids]}]
+  (let [eids-to-retract (d/q '[:find ?s
+                               :in $ ?policy [?statement-id ...]
+                               :where
+                               [?policy :policy/statements ?s]
+                               [?s :statement/sid ?statement-id]]
+                          db [:policy/drn drn] statement-ids)]
+    (map (fn [[eid]] [:db/retractEntity eid]) eids-to-retract)))
+
+(defn add-statements*
+  [db-conn statements-input]
+  (let [{:keys [policy/drn policy/statements]} statements-input
+        retract-sids (into [] (keep :statement/sid) statements)
+        retract-tx (when (seq retract-sids)
+                     [:db.fn/call tx-fn-retract-statements-from-policy
+                      {:policy/drn    drn
+                       :statement-ids retract-sids}])
```

Comment:
We only need to retract statements that have a `sid`, else we assume the user intends to add an unidentified statement to the policy.